poc/fiat_shamir.sage

```diff
@@ -45,48 +45,60 @@ class NIZK:
 
     def prove(self, witness, rng):
         """
-        Proving method using commitment-response format.
-
-        Allows for batching.
+        Proving method using challenge-response format.
         """
         (commitment, challenge, response) = self._prove(witness, rng)
-        # running the verifier here is just a sanity check
         assert self.sigma_protocol.verifier(commitment, challenge, response)
-        return self.sigma_protocol.serialize_commitment(commitment) + self.sigma_protocol.serialize_response(response)
-
+        domain_sep = "NISigmaProtocolShortProof".encode('utf-8')
+        return domain_sep + self.sigma_protocol.serialize_challenge(challenge) + self.sigma_protocol.serialize_response(response)
 
     def verify(self, proof):
-        # Before running the sigma protocol verifier, one must also check that:
-        # - the proof length is exactly commit_bytes_len + response_bytes_len
-        assert len(proof) == self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len, f"Invalid proof length: {len(proof)} != {self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len}"
-        # - proof deserialization successfully produces a valid commitment and a valid response
-        commitment_bytes = proof[:self.sigma_protocol.instance.commit_bytes_len]
-        response_bytes = proof[self.sigma_protocol.instance.commit_bytes_len:]
-        commitment = self.sigma_protocol.deserialize_commitment(commitment_bytes)
+        """
+        Verification method using challenge-response format.
+        """
+        # Check that serialized proof domain separator matches expected value
+        domain_sep = "NISigmaProtocolShortProof".encode('utf-8')
+        domain_sep_bytes = proof[:len(domain_sep)]
+        assert domain_sep_bytes == domain_sep
```

Comment:
While not the case yet, we need to take care that no proof-type tag is a prefix of any other proof-type tag. Otherwise, it's possible to confuse the proof of one type as the proof of a different type. Making each tag the same length would prevent this.